src/Aspire.Cli/Backchannel/AppHostBackchannel.cs

```diff
@@ -88,38 +88,49 @@ await rpc.InvokeWithCancellationAsync(
 
     public async Task ConnectAsync(Process process, string socketPath, CancellationToken cancellationToken)
     {
-        using var activity = _activitySource.StartActivity();
-
-        _process = process;
-
-        if (_rpcTaskCompletionSource.Task.IsCompleted)
+        try
         {
-            throw new InvalidOperationException("Already connected to AppHost backchannel.");
+            using var activity = _activitySource.StartActivity();
+
+            _process = process;
+
+            if (_rpcTaskCompletionSource.Task.IsCompleted)
+            {
+                throw new InvalidOperationException("Already connected to AppHost backchannel.");
+            }
+
+            logger.LogDebug("Connecting to AppHost backchannel at {SocketPath}", socketPath);
+            var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
+            var endpoint = new UnixDomainSocketEndPoint(socketPath);
+            await socket.ConnectAsync(endpoint, cancellationToken);
+            logger.LogDebug("Connected to AppHost backchannel at {SocketPath}", socketPath);
+
+            var stream = new NetworkStream(socket, true);
+            var rpc = JsonRpc.Attach(stream, target);
+
+            var capabilities = await rpc.InvokeWithCancellationAsync<string[]>(
+                "GetCapabilitiesAsync",
+                Array.Empty<object>(),
+                cancellationToken);
+
+            if (!capabilities.Any(s => s == "baseline.v0"))
+            {
+                throw new AppHostIncompatibleException(
+                    $"AppHost is incompatible with the CLI. The AppHost must be updated to a version that supports the baseline.v0 capability.",
+                    "baseline.v0"
+                    );
+            }
+
+            _rpcTaskCompletionSource.SetResult(rpc);
         }
-
-        logger.LogDebug("Connecting to AppHost backchannel at {SocketPath}", socketPath);
-        var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
-        var endpoint = new UnixDomainSocketEndPoint(socketPath);
-        await socket.ConnectAsync(endpoint, cancellationToken);
-        logger.LogDebug("Connected to AppHost backchannel at {SocketPath}", socketPath);
-
-        var stream = new NetworkStream(socket, true);
-        var rpc = JsonRpc.Attach(stream, target);
-
-        var capabilities = await rpc.InvokeWithCancellationAsync<string[]>(
-            "GetCapabilitiesAsync",
-            Array.Empty<object>(),
-            cancellationToken);
-
-        if (!capabilities.Any(s => s == "baseline.v0"))
+        catch (RemoteMethodNotFoundException ex)
         {
+            logger.LogError(ex, "Failed to connect to AppHost backchannel. The AppHost must be updated to a version that supports the baseline.v0 capability.");
```

Comment:
nit, wouldn't this only happen if GetCapabilitiesAsync wasn't defined, which it will be in every version.

Reply:
Correct :) But we are being defensive because this caught us out.